src/Native/System.Private.CoreLib.Native/pal_threading.h

```diff
@@ -0,0 +1,155 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#pragma once
+
+#include "pal_common.h"
+
+#include <stdlib.h>
+
+#include <pthread.h>
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// LowLevelMutex
+
+// Wraps a non-recursive mutex
+class LowLevelMutex
+{
+protected:
+    pthread_mutex_t m_mutex;
+
+#if DEBUG
+private:
+    bool m_isLocked;
+#endif
+
+public:
+    LowLevelMutex(bool abortOnFailure, bool *successRef)
+#if DEBUG
+        : m_isLocked(false)
+#endif
+    {
+        assert(abortOnFailure || successRef != nullptr);
+
+        int error = pthread_mutex_init(&m_mutex, nullptr);
+        if (error != 0)
+        {
+            if (abortOnFailure)
+            {
+                abort();
+            }
+            *successRef = false;
+            return;
+        }
+
+        if (successRef != nullptr)
+        {
+            *successRef = true;
+        }
+    }
+
+    ~LowLevelMutex()
+    {
+        int error = pthread_mutex_destroy(&m_mutex);
+        assert(error == 0);
+
+        UnusedInRelease(error);
+    }
+
+protected:
+    void SetIsLocked(bool isLocked)
+    {
+#if DEBUG
+        assert(m_isLocked != isLocked);
+        m_isLocked = isLocked;
+#endif
+    }
+
+public:
+    void Acquire()
+    {
+        int error = pthread_mutex_lock(&m_mutex);
+        assert(error == 0);
```

Comment:
Shouldn't we be more aggressive in case an error is returned by the OS (e.g., do fail-fast)?